File: go/progress_test.go
package spec

import (
	"encoding/json"
	"testing"
)

func TestToolCallProgressPart_RoundTrip(t *testing.T) {
	part := NewToolCallProgressPart(ToolCallProgressPayload{
		BarID: "stqdm_0", Desc: "Scoring", N: 42, Total: 100, ElapsedS: 12.4, Rate: 3.4, EtaS: 17.1,
	})
	if part.Type() != PartDynamic {
		t.Fatalf("type = %q, want dynamic", part.Type())
	}

	d, ok := part.AsDynamic()
	if !ok || d.Kind != DynamicToolCallProgress {
		t.Fatalf("AsDynamic kind = %q, ok=%v", d.Kind, ok)
	}

	pl, ok := part.AsToolCallProgress()
	if !ok {
		t.Fatal("AsToolCallProgress returned false")
	}
	if pl.BarID != "stqdm_0" || pl.N != 42 || pl.Total != 100 {
		t.Fatalf("payload mismatch: %+v", pl)
	}
	if pl.Status != ProgressRunning {
		t.Fatalf("status = %q, want running (default)", pl.Status)
	}
	if pl.Unit != "it" {
		t.Fatalf("unit = %q, want it (default)", pl.Unit)
	}

	// Verbatim JSON round-trip through the raw-preserving ContentPart.
	raw, err := json.Marshal(part)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var back ContentPart
	if err := json.Unmarshal(raw, &back); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if pl2, ok := back.AsToolCallProgress(); !ok || pl2.EtaS != 17.1 {
		t.Fatalf("round-trip lost payload: ok=%v pl=%+v", ok, pl2)
	}
}

func TestAsToolCallProgress_WrongKind(t *testing.T) {
	other, err := NewDynamicPart(DynamicJSX, map[string]string{"x": "y"}, false)
	if err != nil {
		t.Fatalf("NewDynamicPart: %v", err)
	}
	if _, ok := other.AsToolCallProgress(); ok {
		t.Fatal("AsToolCallProgress must be false for a non-progress dynamic kind")
	}
}

File: typescript/test/progress.test.ts
/**
 * Tests for the dynamic tool_call_progress part (live progress bar).
 *
 * Mirrors go/progress_test.go and python/tests/test_progress.py.
 */
import {describe, expect, it} from 'vitest';
import {
    DYNAMIC_TOOL_CALL_PROGRESS,
    makeToolCallProgressPart,
    type ToolCallProgressPayload,
} from '../src/index';

describe('tool_call_progress dynamic part', () => {
    it('builds a dynamic part with status/unit defaults', () => {
        const part = makeToolCallProgressPart({bar_id: 'stqdm_0', desc: 'Scoring', n: 42, total: 100});
        expect(part.type).toBe('dynamic');
        expect(part.kind).toBe(DYNAMIC_TOOL_CALL_PROGRESS);
        expect(part.interactive).toBe(false);
        const pl = part.payload as ToolCallProgressPayload;
        expect(pl.bar_id).toBe('stqdm_0');
        expect(pl.n).toBe(42);
        expect(pl.total).toBe(100);
        expect(pl.status).toBe('running');
        expect(pl.unit).toBe('it');
    });

    it('lets an explicit status/unit override the default', () => {
        const part = makeToolCallProgressPart({bar_id: 'b', n: 10, total: 10, status: 'done', unit: 'file'});
        const pl = part.payload as ToolCallProgressPayload;
        expect(pl.status).toBe('done');
        expect(pl.unit).toBe('file');
    });
});
